pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +55,83 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+type HeadersLazyLoaded struct {
+	Path   string
+	UserID string
+}
+
+type HeadersLazyLoadedTrackerState struct {
+	LazyLoadedBlocks map[string]int64 `json:"lazy_loaded_blocks"`
+	UserID           string           `json:"user_id"`
+}
+
+func (p *ReaderPool) persist(state HeadersLazyLoadedTrackerState, finalPath string) error {
+	// Create temporary path for fsync
+	tmpPath, err := os.CreateTemp("", "lazy-loaded")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpPath.Name())
+
+	data, err := json.MarshalIndent(state, "", "")
```

Comment:
Should we just use `json.Marshal` here? Not sure we need the file to be pretty-printed (and I'm not sure it will be pretty-printed if both arguments are `""` anyway).